tests/virtualization/universal/waitfor_guests.pm

```diff
@@ -19,17 +19,112 @@ use warnings;
+
+    } elsif (script_run("grep machine '$xml' | grep 'q35'") == 0) {
+        # On q35 add a pcie-root-port and a pcie-to-pci bridge
+
+        # Skip if a pcie-to-pci-bridge is already present
+        return if (script_run("cat $xml | grep 'controller' | grep 'pcie-to-pci-bridge'") == 0);
+
+        my $regex     = '<controller type=.pci. .*model=.pcie-root..*>';
+        my $addbefore = '<controller type="pci" model="pcie-root-port"/><controller type="pci" model="pcie-to-pci-bridge"/>';
+        # the add_before.sh script inserts a given line before the line matching the given regex
+        assert_script_run("virsh dumpxml $guest | /root/add_before.sh '$regex' '$addbefore' > $xml");
+        upload_logs("$xml");
+
+        # Check if settings are applied correctly in the xml
+        assert_script_run("cat $xml | grep 'controller' | grep 'pcie-root-port'");
+        assert_script_run("cat $xml | grep 'controller' | grep 'pcie-to-pci-bridge'");
+
+        # Apply xml settings to VM. Note: They will be applied after reboot.
+        assert_script_run("virt-xml-validate $xml");
+        assert_script_run("virsh define $xml");
+    } else {
+        my $msg = "Unknown machine type";
+        record_soft_failure($msg);
+    }
+}
+
 sub run {
     # Fill the current pairs of hostname & address into /etc/hosts file
     assert_script_run 'virsh list --all';
     add_guest_to_hosts $_, $virt_autotest::common::guests{$_}->{ip} foreach (keys %virt_autotest::common::guests);
     assert_script_run "cat /etc/hosts";
 
+    # Wait for guests to finish installation
+    assert_script_run("wait", timeout => 1200);
+    #script_retry("! ps x | grep -v 'grep' | grep 'virt-install'", retry => 120, delay => 10);
+    # Unfortunately this doesn't cover the second step of the installation, where ssh is available
+    script_run("sleep 300", timeout => 360);    # XXX No idea yet how to prevent this sleep :-(
+    record_info("guests installed", "Guest installation completed");
+    wait_all_guests_online(120);
+
+    # Adding the PCI bridges requires the guests to be shutdown
+    record_info("shutdown guests", "Shutting down all guests");
+    shutdown_guests();
+
+    # Download helper script for adding PCI bridge. add_before adds a line before a given regex.
+    # Usage: add_before REGEX LINE_TO_ADD
+    assert_script_run('curl -v -o /root/add_before.sh ' . data_url('virtualization/add_before.sh'));
+    assert_script_run('chmod 0755 /root/add_before.sh');
+    ## Add a PCIe root port and a PCIe to PCI bridge for hotplugging
+    add_pci_bridge("$_") foreach (keys %virt_autotest::common::guests);
+
+    ## Ensure the reboot policy is set to 'restart'. This needs to happen on shutdown guest
+    # Do a shutdown and start here because some guests might not reboot because of the on_reboot=destroy policy
+    ensure_reboot_policy("$_") foreach (keys %virt_autotest::common::guests);
+    foreach my $guest (keys %virt_autotest::common::guests) {
```

Comment:
Don't you wanna create `dump_xml();` function in `lib/virt_autotest/utils.pm` as well?

Reply:
Good idea! I created `upload_machine_definitions()` now for it, but would keep in in waitfor_guests.pm until it's needed. I would not expect this routine to be used elsewhere unless @guoxuguang or @alice-suse think this would be useful for them as well. It looks like the following

    # Upload xml definitions of all guests
    sub upload_machine_definitions {
        foreach my $guest (keys %virt_autotest::common::guests) {
            my $xml = "$guest.xml";
            assert_script_run("virsh dumpxml $guest | tee $xml");
            upload_logs("$xml");
        }
    }

Reply:
@grisu48 Thanks for being considerate. Pity that product VT QA side have uploaded the guest xml definitions in similar way but not exactly the same way. But  it is a good idea to make it a common lib function.

Reply:
I would vote for leaving it as it is for now, as this is really a simple auxiliary function.